gobblin-core/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,476 @@
+    private final GroupType parquetSchema;
+    private final RecordType recordType;
+
+    public enum RecordType {
+      ROOT, CHILD
+    }
+
+    public RecordConverter(JsonSchema recordSchema, WorkUnitState state) {
+      this(recordSchema, state, CHILD);
+    }
+
+    public RecordConverter(JsonSchema recordSchema, WorkUnitState state, RecordType recordType) {
+      super(recordSchema);
+      this.converters = new HashMap<>();
+      this.recordType = recordType;
+      this.parquetSchema = buildSchema(recordSchema.getDataTypeValues(), state);
+    }
+
+    @Override
+    Object convertField(JsonElement value) {
+      ParquetGroup r1 = new ParquetGroup(parquetSchema);
+      JsonObject inputRecord = value.getAsJsonObject();
+      for (Map.Entry<String, JsonElement> entry : inputRecord.entrySet()) {
+        JsonElementConverter converter = this.converters.get(entry.getKey());
+        r1.add(entry.getKey(), converter.convert(entry.getValue()));
+      }
+      return r1;
+    }
+
+    @Override
+    public GroupType schema() {
+      return parquetSchema;
+    }
+
+    private GroupType buildSchema(JsonArray inputSchema, WorkUnitState workUnit) {
+      List<Type> parquetTypes = new ArrayList<>();
+      for (JsonElement element : inputSchema) {
+        JsonObject map = (JsonObject) element;
+        JsonSchema elementSchema = new JsonSchema(map);
+        String columnName = elementSchema.getColumnName();
+        JsonElementConverter converter = JsonElementConversionFactory.getConverter(elementSchema, workUnit, false);
+        Type schemaType = converter.schema();
+        this.converters.put(columnName, converter);
+        parquetTypes.add(schemaType);
+      }
+      String docName = this.jsonSchema.getColumnName();
+      switch (recordType) {
+        case ROOT:
+          return new MessageType(docName, parquetTypes);
+        case CHILD:
+          return new GroupType(this.jsonSchema.optionalOrRequired(), docName, parquetTypes);
+        default:
+          throw new RuntimeException("Unsupported Record type");
+      }
+    }
+  }
+
+  public static class MapConverter extends CollectionConverter {
+
+    private final WorkUnitState state;
+
+    public MapConverter(JsonSchema mapSchema, WorkUnitState state) {
+      super(mapSchema, mapSchema.getElementTypeUsingKey(MAP_ITEMS_KEY));
+      this.state = state;
+    }
+
+    @Override
+    Object convertField(JsonElement value) {
+      ParquetGroup mapGroup = new ParquetGroup(schema());
+      JsonElementConverter converter = getElementConverter();
+      JsonObject map = (JsonObject) value;
+
+      for (Map.Entry<String, JsonElement> entry : map.entrySet()) {
+        ParquetGroup entrySet = (ParquetGroup) mapGroup.addGroup(MAP_KEY);
+        entrySet.add(MAP_KEY_COLUMN_NAME, entry.getKey());
+        entrySet.add(MAP_VALUE_COLUMN_NAME, converter.convert(entry.getValue()));
+      }
+
+      return mapGroup;
+    }
+
+    @Override
+    public GroupType schema() {
```

Comment:
The schema can be computed on initialization and return it here. In this way, every call to `schema()` won't trigger duplicate computation.